book/src/dev/rfcs/0002-parallel-verification.md

```diff
@@ -0,0 +1,324 @@
+    complete, assumes that the `previous_block_hash` fields of these blocks
+    form an unbroken chain from checkpoint to checkpoint, and starts
+    processing the checkpoint range. (This constraint is an implicit part of
+    the `CheckpointVerifier` design.)
+  * `process_checkpoint_range`: makes sure that the blocks in the checkpoint
+    range have an unbroken chain of previous block hashes.
+* **State Updates:**
+  * *As Above:* the `CheckpointVerifier` returns success to the `ChainVerifier`,
+    which sends verified `Block`s to the state service.
+  * The state service implements the chain order consensus rule, which makes sure
+    each block is added to state after its previous block.
+
+### Block Verification
+[block-verification]: #block-verification
+
+The `BlockVerifier` performs detailed verification of recent blocks, in parallel.
+
+Here is how the `BlockVerifier` implements each verification stage:
+
+* **Structural Verification:**
+  * *As Above:* the `BlockVerifier` accepts parsed `Block` structs.
+* **Semantic Verification:**
+  * *As Above:* verifies each field in the block. Defers any data dependencies as
+    long as possible, awaits those data dependencies, then performs data
+    dependent checks.
+  * Note: Since futures are executed concurrently, we can use the same function
+    to:
+    * perform context-free verification,
+    * perform verification with deferred data dependencies,
+    * await data dependencies, and
+    * check data dependencies.
+    To maximise concurrency, we should write verification functions in this
+    specific order, so the awaits are as late as possible.
+* **State Updates:**
+  * *As Above:* the `BlockVerifier` returns success to the `ChainVerifier`,
+    which sends verified `Block`s to the state service.
+  * The state service implements the chain order consensus rule, which makes sure
+    each block is added to state after its previous block.
+
+## Zcash Protocol Design
+[zcash-protocol]: #zcash-protocol
+
+When designing a change to the Zcash protocol, minimise the data dependencies
+between blocks.
+
+Try to create designs that:
+* Eliminate data dependencies,
+* Make the changes depend on a version field in the block header or transaction,
+* Make the changes depend on the current Network Upgrade,
+* Make the changes depend on a field in the current block, with an additional
+  consensus rule to check that field against previous blocks, or
+* Prefer dependencies on older blocks, rather than newer blocks
+  (older blocks are more likely to verify earlier).
+
+Older dependencies have a design tradeoff:
+* depending on multiple blocks is complex,
+* depending on the previous block makes parallel verification harder.
+
+When making decisions about this dependency tradeoff, consider:
+* how the data dependency could be deferred, and
+* the CPU cost of the verification - if it is trivial, then it does not matter if
+  the verification is parallelised.
```

Comment:
I agree that efficient verification and simple data dependencies are a few goals, among many important goals.

But I think that mentioning the other goals for protocol updates is out of scope for this RFC. I also don't think that prioritising those goals is in scope for this RFC.

How can I make that clearer in this RFC?
Do you have a specific wording change you'd like to see?
Is there any wording that makes it seem like a requirement, rather than a design consideration?

Reply:
my vote would be to remove this section entirely.

Reply:
I don't like "Prefer dependencies on older blocks, rather than newer blocks"; it seems unlikely that interesting changes could satisfy that, anyway. All of the rest seem like sensible preferences that I'd be happy to consider for any future protocol design (if ZF engineers were to argue for them then).

Reply:
Thanks for the review!

When I said "depend on older blocks", I was mainly thinking of the UTXO set, where we depend on data from some older blocks, but potentially also data up to the parent block. In particular, the coinbase maturity rule ensures that coinbase UTXO transactions depend on blocks that are at least 100 blocks behind the tip.

But I think you're right, it's unlikely that we will create rules that create dependencies on older blocks. And in most cases, there will be a security or usability impact, which is more important than data dependencies.

So I'll delete that particular guideline.

Reply:
Deleted in 21bb69ad58388df93db224324fd8116479375a4d.